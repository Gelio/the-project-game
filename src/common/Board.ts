import { BoardSize } from '../interfaces/BoardSize';

export class Board {
  public readonly size: BoardSize;

  constructor(size: BoardSize) {
    this.size = size;
  }

  public reset() {
    // TODO: implement this
  }
}
