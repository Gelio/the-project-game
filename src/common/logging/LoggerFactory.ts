import { Stream } from 'stream';
import * as winston from 'winston';

const typelessWinston: any = winston;

export class LoggerFactory {
  public logLevel: winston.NPMLoggingLevel = 'silly';

  public createLogger(transports: winston.TransportInstance[]): winston.LoggerInstance {
    const format = typelessWinston.format;

    return typelessWinston.createLogger({
      format: format.timestamp(),
      transports,
      level: this.logLevel
    });
  }

  public createConsoleLogger(): winston.LoggerInstance {
    const format = typelessWinston.format;

    const consoleTransport = new winston.transports.Console(<any>{
      format: format.combine(
        format.colorize(),
        format.align(),
        format.printf((info: any) => `${info.timestamp} ${info.level}: ${info.message}`)
      )
    });

    return this.createLogger([consoleTransport]);
  }

  public createEmptyLogger(): winston.LoggerInstance {
    const emptyStreamTransport = new typelessWinston.transports.Stream({
      silent: true,
      stream: new Stream.Writable({ objectMode: true, write: () => null })
    });

    return this.createLogger([emptyStreamTransport]);
  }
}
