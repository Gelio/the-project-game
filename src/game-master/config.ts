export const config = {
  uiLabelStyle: '{black-fg}{white-bg}{bold}',

  tests: {
    timestampDifferenceThreshold: 5
  }
};
