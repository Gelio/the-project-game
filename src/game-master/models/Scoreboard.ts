export class Scoreboard {
  public team1Score: number;
  public team2Score: number;
  public readonly scoreLimit: number;

  constructor(scoreLimit: number) {
    this.scoreLimit = scoreLimit;
    this.team1Score = 0;
    this.team2Score = 0;
  }
}
