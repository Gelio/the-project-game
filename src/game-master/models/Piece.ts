import { Point } from '../../common/Point';

export class Piece {
  public isSham: boolean;

  /**
   * When the piece is picked up, this is the position of the player
   */
  public position: Point;
  public isPickedUp: boolean;
}
