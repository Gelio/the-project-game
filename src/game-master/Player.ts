import { Point } from '../common/Point';
import { TeamId } from '../common/TeamId';
import { Piece } from './models/Piece';

export class Player {
  public isConnected: boolean;
  public teamId: TeamId;
  public isLeader: boolean;
  public isBusy = false;

  public playerId: number;
  public position: Point | null = null;
  public heldPiece: Piece | null = null;
}
