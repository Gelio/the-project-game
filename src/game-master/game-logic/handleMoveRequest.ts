import { createDelay } from '../../common/createDelay';
import { Point } from '../../common/Point';

import { MoveRequest } from '../../interfaces/requests/MoveRequest';
import { MoveResponse } from '../../interfaces/responses/MoveResponse';

import { Player } from '../Player';
import { ProcessMessageResult } from '../ProcessMessageResult';

import { MessageHandlerDependencies } from './MessageHandlerDependencies';

import { Direction } from '../../interfaces/Direction';

export function handleMoveRequest(
  { board, actionDelays }: MessageHandlerDependencies,
  sender: Player,
  _moveRequest: MoveRequest
): ProcessMessageResult<MoveResponse> {
  const playerPosition = sender.position;
  if (!playerPosition) {
    return {
      valid: false,
      reason: 'Invalid player position. Something is wrong with GM'
    };
  }

  let newPosition: Point = new Point(-1, -1);

  switch (_moveRequest.payload.direction) {
    case Direction.Down: {
      newPosition = new Point(playerPosition.x, playerPosition.y + 1);
      break;
    }
    case Direction.Up: {
      newPosition = new Point(playerPosition.x, playerPosition.y - 1);
      break;
    }
    case Direction.Left: {
      newPosition = new Point(playerPosition.x - 1, playerPosition.y);
      break;
    }
    case Direction.Right: {
      newPosition = new Point(playerPosition.x + 1, playerPosition.y);
      break;
    }
    default:
  }
  try {
    board.movePlayer(sender, newPosition);
  } catch (error) {
    return {
      valid: false,
      reason: error
    };
  }
  const responsePromise = createDelay(actionDelays.move).then((): MoveResponse => ({
    type: 'MOVE_RESPONSE',
    payload: {
      timestamp: Date.now(),
      distanceToPiece: board.getDistanceToClosestPiece(newPosition)
    },
    recipientId: sender.playerId,
    senderId: -1
  }));

  return {
    valid: true,
    delay: actionDelays.move,
    responseMessage: responsePromise
  };
}
