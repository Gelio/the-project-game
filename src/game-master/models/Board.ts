import { Point } from '../../common/Point';
import { BoardSize } from '../../interfaces/BoardSize';
import { Player } from '../Player';
import { Piece } from './Piece';
import { Tile } from './tiles/Tile';

export class Board {
  public readonly size: BoardSize;
  public readonly tiles: Tile[][];
  public readonly pieces: Piece[] = [];

  constructor(size: BoardSize, tiles: Tile[][]) {
    this.size = size;
    this.tiles = tiles;
  }

  public reset() {
    // TODO: implement this
  }

  public getTileAtPosition(position: Point) {
    const tileColumn = this.tiles[position.x];
    if (!tileColumn) {
      throw new Error('Invalid X coordinate');
    }

    const tile = tileColumn[position.y];
    if (!tile) {
      throw new Error('Invalid Y coordinate');
    }

    return tile;
  }

  public addPlayer(player: Player) {
    this.getTileAtPosition(player.position).player = player;
  }

  public removePlayer(point: Point) {
    this.tiles[point.x][point.y].player = null;
  }

  public movePlayer(player: Player, newPosition: Point) {
    const previousTile = this.getTileAtPosition(player.position);
    const newTile = this.getTileAtPosition(newPosition);

    if (previousTile.player !== player) {
      throw new Error('Old player position corrupted');
    }

    if (newTile.player) {
      throw new Error('Two players cannot stand on the same tile');
    }

    previousTile.player = null;
    newTile.player = player;
  }

  public addPiece(piece: Piece) {
    if (this.pieces.indexOf(piece) !== -1) {
      throw new Error('Piece already added');
    }

    const tile = this.getTileAtPosition(piece.position);
    if (tile.piece) {
      throw new Error('Piece already exists at that position');
    }

    tile.piece = piece;
    this.pieces.push(piece);
  }

  public removePiece(piece: Piece) {
    const index = this.pieces.indexOf(piece);
    if (index === -1) {
      throw new Error('Piece has not been added to the game previously');
    }

    this.pieces.splice(index, 1);

    if (!piece.isPickedUp) {
      const tile = this.getTileAtPosition(piece.position);
      tile.piece = null;
    }
  }

  public movePiece(piece: Piece, newPosition: Point) {
    const index = this.pieces.indexOf(piece);
    if (index === -1) {
      throw new Error('Piece has not been added to the game previously');
    }

    const previousTile = this.getTileAtPosition(piece.position);
    const newTile = this.getTileAtPosition(newPosition);

    if (previousTile.piece !== piece) {
      throw new Error('Old piece position corrupted');
    }

    if (newTile.piece) {
      throw new Error('Cannot move a piece on a tile which already has one');
    }

    previousTile.piece = null;
    newTile.piece = piece;
    piece.position = newPosition;
  }
}
