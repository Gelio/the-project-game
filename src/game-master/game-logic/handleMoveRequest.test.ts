import { LoggerInstance } from 'winston';

import { LoggerFactory } from '../../common/logging/LoggerFactory';
import { Point } from '../../common/Point';

import { ActionDelays } from '../../interfaces/ActionDelays';
import { Direction } from '../../interfaces/Direction';
import { MoveResponse } from '../../interfaces/responses/MoveResponse';

import { Board } from '../models/Board';
import { Piece } from '../models/Piece';
import { Scoreboard } from '../models/Scoreboard';

import { config } from '../config';
import { Player } from '../Player';
import {
  InvalidMessageResult,
  ProcessMessageResult,
  ValidMessageResult
} from '../ProcessMessageResult';

import { handleMoveRequest } from './handleMoveRequest';

describe('[GM] handleMoveRequest ', () => {
  let board: Board;
  let actionDelays: ActionDelays;
  let player: Player;
  let logger: LoggerInstance;
  let scoreboard: Scoreboard;

  beforeEach(() => {
    board = new Board(
      {
        goalArea: 5,
        taskArea: 20,
        x: 10
      },
      10
    );

    actionDelays = <any>{
      move: 500
    };

    scoreboard = new Scoreboard(10);

    player = new Player();
    player.isBusy = true;
    player.playerId = 'player1';
    board.addPlayer(player);
    board.movePlayer(player, new Point(5, 15));

    const loggerFactory = new LoggerFactory();
    loggerFactory.logLevel = 'error';

    logger = loggerFactory.createEmptyLogger();

    jest.useFakeTimers();
  });

  afterEach(() => {
    jest.useRealTimers();
  });

  function executeHandler(direction: Direction): ProcessMessageResult<MoveResponse> {
    return <any>handleMoveRequest(
      {
        board,
        playersContainer: <any>{},
        actionDelays: <any>actionDelays,
        logger,
        scoreboard
      },
      player,
      {
        senderId: 'player1',
        payload: {
          direction
        },
        type: 'MOVE_REQUEST'
      }
    );
  }

  it('should always be valid when move is valid', () => {
    const result = executeHandler(Direction.Down);

    expect(result.valid).toBe(true);
  });

  it('should reject request when move is invalid', async () => {
    board.movePlayer(player, new Point(0, 0));
    const result = executeHandler(Direction.Up);

    expect(result.valid).toBe(false);
  });

  it('should return the response after delay if move is valid', () => {
    const result = <ValidMessageResult<MoveResponse>>executeHandler(Direction.Down);

    expect(result.delay).toBe(actionDelays.move);

    result.responseMessage.then(response => {
      expect(response.recipientId).toBe(player.playerId);
    });

    expect.assertions(2);
    jest.advanceTimersByTime(actionDelays.move);
  });

  it('should move player down', async () => {
    executeHandler(Direction.Down);

    expect((<Point>player.position).x).toBe(5);
    expect((<Point>player.position).y).toBe(16);
  });

  it('should move player up', async () => {
    executeHandler(Direction.Up);

    expect((<Point>player.position).x).toBe(5);
    expect((<Point>player.position).y).toBe(14);
  });

  it('should move player right', async () => {
    executeHandler(Direction.Right);

    expect((<Point>player.position).x).toBe(6);
    expect((<Point>player.position).y).toBe(15);
  });

  it('should move player left', async () => {
    executeHandler(Direction.Left);

    expect((<Point>player.position).x).toBe(4);
    expect((<Point>player.position).y).toBe(15);
  });

  it('should reject request when direction is invalid', () => {
    const result = executeHandler(<any>'test');

    expect(result.valid).toBe(false);
    expect((<InvalidMessageResult>result).reason).toMatchSnapshot();
  });

  it("should reject request when player tries to move into other team's area", () => {
    board.movePlayer(player, new Point(0, board.size.goalArea + board.size.taskArea));

    const result = executeHandler(Direction.Down);

    expect(result.valid).toBe(false);
    expect((<InvalidMessageResult>result).reason).toMatchSnapshot();
  });

  it("should reject request when board can't move player to new position", () => {
    player.position = new Point(6, 6);

    const result = executeHandler(Direction.Down);

    expect(result.valid).toBe(false);
    expect((<InvalidMessageResult>result).reason).toMatchSnapshot();
  });

  describe('response', () => {
    it('should contain current timestamp', async () => {
      const result = <ValidMessageResult<MoveResponse>>executeHandler(Direction.Down);
      jest.advanceTimersByTime(actionDelays.move);

      const { payload } = await result.responseMessage;
      expect(Math.abs(payload.timestamp - Date.now())).toBeLessThanOrEqual(
        config.tests.timestampDifferenceThreshold
      );
    });

    it('should contain distance to closest piece', async () => {
      const piece = new Piece();

      piece.position = (<Point>player.position).clone();
      board.addPiece(piece);

      const result = <ValidMessageResult<MoveResponse>>executeHandler(Direction.Down);
      jest.advanceTimersByTime(actionDelays.move);

      const { payload } = await result.responseMessage;
      expect(payload.distanceToPiece).toBe(1);
    });

    it('should have distance to the closest piece set to -1 when there are no pieces on the board', async () => {
      const result = <ValidMessageResult<MoveResponse>>executeHandler(Direction.Down);
      jest.advanceTimersByTime(actionDelays.move);

      const { payload } = await result.responseMessage;

      expect(payload.distanceToPiece).toBe(-1);
    });

    it("should contain valid distances to closest piece (when it's far away)", async () => {
      const piece = new Piece();
      piece.isPickedUp = false;
      piece.isSham = true;
      piece.position = new Point(5, 5);
      board.addPiece(piece);

      const result = <ValidMessageResult<MoveResponse>>executeHandler(Direction.Down);
      jest.advanceTimersByTime(actionDelays.move);

      const { payload } = await result.responseMessage;

      expect(payload.distanceToPiece).toBe(11);
    });
  });
});
