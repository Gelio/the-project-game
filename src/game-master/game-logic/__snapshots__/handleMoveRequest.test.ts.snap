// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[GM] handleMoveRequest  should reject request when direction is invalid 1`] = `"Invalid direction."`;
